pkg/controller/sync/aggregator/aggregator.go

```diff
@@ -0,0 +1,105 @@
+package aggregator
+
+import (
+	"fmt"
+	gosync "sync"
+
+	"golang.org/x/exp/maps"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type KindName struct {
+	Kind string
+	Name string
+}
+
+func NewGVKAggregator(allowedKeyKinds []string) *GVKAgreggator {
+	return &GVKAgreggator{
+		store:        make(map[KindName]map[schema.GroupVersionKind]struct{}),
+		reverseStore: make(map[schema.GroupVersionKind]map[KindName]struct{}),
+
+		allowedKeyKinds: allowedKeyKinds,
+	}
+}
+
+// todo comments.
+type GVKAgreggator struct {
```

Comment:
This type will need an `IsSynced(gvk) bool` or similar function so that constraint templates can know whether the required GVKs are being synced.

Reply:
There was an initial plan to have `IsReferenced` or similar for a gvk here. The idea was to use ListGVKs() instead of `IsSynced` but I reckon we can use `IsSynced` instead.